### PR TITLE
Download project devfile

### PR DESCRIPTION
**Pull request by EnriqueL8**
**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:
This PR adds a new flag `--downloadSource` which pulls down the first project specified in the list of project in the devfile. 

**Which issue(s) this PR fixes**:
Implements #2582

**How to test changes / Special notes to the reviewer**:
Added unit and integrations testing for this feature.
NOTE: We add a second context folder for each test, because we need the folder to either be empty or only contain a devfile.yaml. If use only one context, we have the config.yaml file present.

### CODE REVIEW

**Comment by kadel** [comment]
a lot of test failures :-(

```
/home/travis/gopath/bin/gosec
gosec -severity medium -confidence medium -exclude G304,G204 -quiet  ./...
Results:
[/home/travis/gopath/src/github.com/openshift/odo/pkg/util/util.go:895] - G110 (CWE-409): Potential DoS vulnerability via decompression bomb (Confidence: MEDIUM, Severity: MEDIUM)
  > io.Copy(outFile, rc)
[/home/travis/gopath/src/github.com/openshift/odo/tests/helper/helper_filesystem.go:244] - G306 (CWE-): Expect WriteFile permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
  > ioutil.WriteFile(devfileLocation, []byte(output), 0644)

```

If it makes sense to ignore the security check you can add a comment like this `// #nosec G306` before the line to ignore the security check. But we need to be careful about that.

**Comment by kadel** [comment]
It would be great to have some integration tests that will execute the command with the flag and verify that the files get really downloaded

**Comment by EnriqueL8** [comment]
Seems like we can change the file permission to pass that security check. We added a series of integration tests for each type of project with the flag.

**Comment by johnmcollier** [comment]
@EnriqueL8 Travis build failing due to the following errors
```
golangci-lint run ./... --timeout 5m

pkg/odo/cli/component/create.go:721:3: ineffectual assignment to `err` (ineffassign)

		err = errors.Errorf("Project type not supported")

		^

pkg/util/util.go:827:4: Error return value is not checked (errcheck)

		}()

		 ^

pkg/util/util.go:876:15: Error return value of `os.MkdirAll` is not checked (errcheck)

			os.MkdirAll(fpath, os.ModePerm)

			           ^

pkg/util/util.go:833:9: SA9003: empty branch (staticcheck)

	} else {
```

Can you add error checks to those lines?

**Comment by EnriqueL8** [comment]
@johnmcollier pushed changes for handling errors, but still getting some weird test failures on travis with something not related to my changes.

File: tests/integration/devfile/cmd_devfile_create_test.go [hunk 1]
```diff
@@ -114,4 +115,60 @@ var _ = Describe("odo devfile create command tests", func() {
...
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", devfile), filepath.Join(contextDevfile, devfile))
+			err := helper.ReplaceDevfileField(devfile, "location", "https://github.com/che-samples/web-nodejs-sample/archive/master.zip")
+			if err != nil {
+				log.Error("Could not replace the entry in the devfile: " + err.Error())
+			}
+			err = helper.ReplaceDevfileField(devfile, "type", "zip")
+			if err != nil {
+				log.Error("Could not replace the entry in the devfile: " + err.Error())
+			}
+			helper.CmdShouldPass("odo", "create", "nodejs", "--downloadSource", "--devfile", devfile, "--context", filepath.Join(context, "config.yaml"))
```
Comment (GeekArthur) [inline tests/integration/devfile/cmd_devfile_create_test.go:168]: FYI, currently we only support `--context` for s2i component, `--context` support for devfile component will be tacked by this issue https://github.com/openshift/odo/issues/2824. For we now we can only use helper.Chdir(<destination>) instead of using `--context`

**Comment by johnmcollier** [comment]
@EnriqueL8 Integration tests are failing also due to the `--now` flag being removed from `odo create`

**Comment by johnmcollier** [comment]
Fixed the build issues on my machine, I had to update my Go version to at least 1.13

We need to document this in the repo if odo will now require a higher Go version to build it, as it previously required a much lower Go version

File: pkg/odo/cli/component/create.go [hunk 2]
```diff
@@ -803,12 +889,13 @@ func NewCmdCreate(name, fullName string) *cobra.Command {
 
 	if experimental.IsExperimentalModeEnabled() {
 		componentCreateCmd.Flags().StringVarP(&co.devfileMetadata.componentNamespace, "namespace", "n", "", "Create devfile component under specific namespace")
+		componentCreateCmd.Flags().BoolVar(&co.devfileMetadata.downloadSource, "downloadSource", false, "")
```
Comment (johnmcollier) [inline pkg/odo/cli/component/create.go:892]: Can you add an example in the help text for `odo create` for the `downloadSource` flag?

File: pkg/odo/cli/component/create.go [hunk 3]
```diff
@@ -776,12 +863,13 @@ func NewCmdCreate(name, fullName string) *cobra.Command {
...
+		componentCreateCmd.Flags().StringVar(&co.devfileMetadata.devfilePath, "devfile", "./devfile.yaml", "Path to a devfile.yaml")
```
[the remark this answers is not in the document]
Comment (GeekArthur) [reply pkg/odo/cli/component/create.go:893]: @johnmcollier I don't have PR for #2862, the one you see under #2862 is this PR as we mention this PR in #2862

[the remark this answers is not in the document]
Comment (GeekArthur) [reply pkg/odo/cli/component/create.go:893]: But I leave my comment regarding `--devile` here: https://github.com/openshift/odo/pull/2902#discussion_r410303134

File: tests/integration/devfile/cmd_devfile_create_test.go (hunk 1 above)
Comment (EnriqueL8) [reply tests/integration/devfile/cmd_devfile_create_test.go:168]: The issue is that we don't want to have the `config.yaml` and the `devfile.yaml` in the same folder, but we still need the experimental flag set which is stored in `config.yaml`? So if I don't specify context, we cannot read the preference and if I chdir with the two files my code fails.

File: pkg/util/util.go [hunk 4]
```diff
@@ -717,6 +720,199 @@ func FilterIgnores(filesChanged, filesDeleted, absIgnoreRules []string) (filesCh
...
+
+	if len(files) > 1 {
+		return errors.Errorf("Folder is not empty. It can only contain the devfile used.")
+	} else if len(files) == 1 {
+		file := files[0]
+		if file.IsDir() {
+			return errors.Errorf("Folder is not empty. It contains a subfolder.")
+		}
+		fileName := files[0].Name()
+		if fileName != devfilePath {
```
[the remark this answers is not in the document]
Comment (EnriqueL8) [reply pkg/util/util.go:739]: Good catch!

File: pkg/odo/cli/component/create.go [hunk 5]
```diff
@@ -677,23 +681,105 @@ func (co *CreateOptions) Validate() (err error) {
...
+				err := util.DownloadFile(co.devfileMetadata.devfileRegistry+co.devfileMetadata.devfileLink, DevfilePath)
+				if err != nil {
+					return errors.Wrap(err, "Faile to download devfile.yaml for devfile component")
+				}
+				co.devfileMetadata.devfilePath = DevfilePath
 			}
 
-			if !util.CheckPathExists(DevfilePath) {
-				err = util.DownloadFile(co.devfileMetadata.devfileRegistry+co.devfileMetadata.devfileLink, DevfilePath)
+			if co.devfileMetadata.downloadSource {
```
Comment (johnmcollier) [inline pkg/odo/cli/component/create.go:771]: When we're downloading the source + devfile into an empty directory, it doesn't seem to properly download the devfile:
```
Johns-MacBook-Pro-3:testcreate johncollier$ odo create java-spring-boot --downloadSource true
Experimental mode is enabled, use at your own risk

Validation
 ✓  Checking devfile compatibility [80921ns]
 ✓  Validating devfile component [210766ns]
 ✗  open /Users/johncollier/testcreate/devfile.yaml: no such file or directory
```

File: pkg/odo/cli/component/create.go [hunk 6]
```diff
@@ -776,12 +863,13 @@ func NewCmdCreate(name, fullName string) *cobra.Command {
...
 	if experimental.IsExperimentalModeEnabled() {
 		componentCreateCmd.Flags().StringVarP(&co.devfileMetadata.componentNamespace, "namespace", "n", "", "Create devfile component under specific namespace")
+		componentCreateCmd.Flags().BoolVar(&co.devfileMetadata.downloadSource, "downloadSource", false, "")
+		componentCreateCmd.Flags().StringVar(&co.devfileMetadata.devfilePath, "devfile", "./devfile.yaml", "Path to a devfile.yaml")
 	}
 
 	componentCreateCmd.SetUsageTemplate(odoutil.CmdUsageTemplate)
 
 	// Adding `--now` flag
-	genericclioptions.AddNowFlag(componentCreateCmd, &co.now)
```
[the remark this answers is not in the document]
Comment (EnriqueL8) [reply pkg/odo/cli/component/create.go:811]: Oh this must a been a mistake

File: pkg/odo/cli/component/create.go [hunk 7]
```diff
@@ -803,12 +889,13 @@ func NewCmdCreate(name, fullName string) *cobra.Command {
 
 	if experimental.IsExperimentalModeEnabled() {
 		componentCreateCmd.Flags().StringVarP(&co.devfileMetadata.componentNamespace, "namespace", "n", "", "Create devfile component under specific namespace")
+		componentCreateCmd.Flags().BoolVar(&co.devfileMetadata.downloadSource, "downloadSource", false, "")
+		componentCreateCmd.Flags().StringVar(&co.devfileMetadata.devfilePath, "devfile", "./devfile.yaml", "Path to a devfile.yaml")
```
[the remark this answers is not in the document]
Comment (EnriqueL8) [reply pkg/odo/cli/component/create.go:893]: So how should we proceed? I am not too sure what to do in this case

File: tests/integration/devfile/cmd_devfile_create_test.go (hunk 1 above)
Comment (johnmcollier) [reply tests/integration/devfile/cmd_devfile_create_test.go:168]: Why don't we want to have the config.yaml in the directory? 

FWIW, using the `--context` flag right now in odo for devfiles is pretty much a no-op so it should be fine to leave in.

File: pkg/odo/cli/component/create.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (GeekArthur) [reply pkg/odo/cli/component/create.go:893]: The easiest way is to remove the `--devfile` support in the PR and still uses the default devfile path which is `./devfile.yaml` , then use separate PR to implement the design for `--devfile`

File: tests/integration/devfile/cmd_devfile_create_test.go (hunk 1 above)
Comment (GeekArthur) [reply tests/integration/devfile/cmd_devfile_create_test.go:168]: @EnriqueL8 you can check our devfile push test here https://github.com/openshift/odo/blob/master/tests/integration/devfile/cmd_devfile_push_test.go#L31

- It uses context directory to store `config.yaml` and set up the file location via configuring `GLOBALODOCONFIG` env var
- It creates a new project directory to store both template and `devfile.yaml`

With that change `config.yaml` and `devfile.yaml` are in different directory, and we can use `helper.Chdir(<destination>)` to switch to the  new project directory to do all the devfile related operations and we can read the preference from there as well.

**Comment by GeekArthur** [comment]
Not sure if only I hit this issue, after I pull down the branch and run `odo create --help` the flags section is broken

```
odo create --help                                                                                                                      ✔  13:25:59 
Create a configuration describing a component.

 If a component name is not provided, it'll be auto-generated.

 A full list of component types that can be deployed is available using: 'odo catalog list'

 By default, builder images (component type) will be used from the current namespace. You can explicitly supply a namespace by using: odo create namespace/name:version If version is not specified by default, latest will be chosen as the version.

Usage:
  odo create <component_type> [component_name] [flags]

Examples:
  # Create new Node.js component with the source in current directory.
  odo create nodejs

  # Create new Node.js component named 'frontend' with the source in './frontend' directory
  odo create nodejs frontend --context ./frontend

  # Create new Java component with binary named sample.jar in './target' directory
  odo create java:8  --binary target/sample.jar

  # Create new Node.js component with source from remote git repository
  odo create nodejs --git https://github.com/openshift/nodejs-ex.git

  # Create new Node.js component with custom ports, additional environment variables and memory and cpu limits
  odo create nodejs --port 8080,8100/tcp,9100/udp --env key=value,key1=value1 --memory 4Gi --cpu 2

Flags:
template: top:15:2: executing "top" at <CapitalizeFlagDescriptions .LocalFlags>: error calling CapitalizeFlagDescriptions: runtime error: index out of range [0] with length 0
```

**Comment by johnmcollier** [comment]
@GeekArthur That's from the `--now` flag being accidentally removed. Once @EnriqueL8 adds it back, it should be fixed.

File: tests/integration/devfile/cmd_devfile_create_test.go (hunk 1 above)
Comment (GeekArthur) [reply tests/integration/devfile/cmd_devfile_create_test.go:168]: FYI, I just drop a PR https://github.com/openshift/odo/pull/2917 for `--context` support for devfile component

File: pkg/util/util.go [hunk 8]
```diff
@@ -717,6 +720,196 @@ func FilterIgnores(filesChanged, filesDeleted, absIgnoreRules []string) (filesCh
...
+	remote = strings.Replace(remote, ":", "/", 1)
+	remote = strings.Replace(remote, "git@", "https://", 1)
+	return remote
+}
+
+// GetGitHubZipURL downloads a repo from a URL to a destination
+func GetGitHubZipURL(repoURL string) (string, error) {
+	var url string
+	// Convert ssh remote to https
+	if strings.HasPrefix(repoURL, "git@") {
```
[the remark this answers is not in the document]
Comment (EnriqueL8) [reply pkg/util/util.go:758]: I believe we would still be able to use http, but the user will need to provide a token for authentication.

**Comment by EnriqueL8** [comment]
@johnmcollier @GeekArthur Updated with requested changes. Had to comment some integration tests out due to recent changes from `https://github.com/openshift/odo/pull/2798`, and there is no interactive support yet for testing. Once  https://github.com/openshift/odo/issues/2194 is implemented, those tests can be modified and added.

**Comment by GeekArthur** [comment]
> @johnmcollier @GeekArthur Updated with requested changes. Had to comment some integration tests out due to recent changes from `https://github.com/openshift/odo/pull/2798`, and there is no interactive support yet for testing. Once #2194 is implemented, those tests can be modified and added.

Another way for testing pre-defined devfile is via configuring registry to let the configured registry hosts the pre-defined devfile, once this issue https://github.com/openshift/odo/issues/2635 is done, we can go that route for testing.

**Comment by kadel** [comment]
> Changes look good, pulled down the branch and gave it a try myself.
> 
> **But**, I'm not sure how we should communicate / handle the requirement that Go 1.13 (or higher) is now needed to build odo. CCing @kadel for discussion.

We are already using Go 1.13 in all our builds (CI/Release etc..)
It would be good to state minimal Go version requirements in https://github.com/openshift/odo/blob/master/docs/dev/development.adoc
